batch2/batch/batch.py

```diff
@@ -1,19 +1,1278 @@
+            await pod_changed(pod)
+        except Exception as exc:
+            log.exception(f'driver event loop failed due to exception: {exc}')
+
+
+async def polling_event_loop():
+    await asyncio.sleep(1)
+    while True:
+        try:
+            await refresh_pods()
+        except Exception as exc:
+            log.exception(f'polling event loop failed due to exception: {exc}')
+        await asyncio.sleep(60 * 10)
+
+
+async def db_cleanup_event_loop():
+    await asyncio.sleep(1)
+    while True:
+        try:
+            for record in await db.batch.get_finished_deleted_records():
+                batch = Batch.from_record(record, deleted=True)
+                await batch.delete()
+        except Exception as exc:  # pylint: disable=W0703
+            log.exception(f'Could not delete batches due to exception: {exc}')
+        await asyncio.sleep(REFRESH_INTERVAL_IN_SECONDS)
+
+
+async def profile_loop():
+    await asyncio.sleep(1)
+    while True:
+        app['profile'].disable()
+        app['profile'].print_stats(sort='time')
+        app['profile'].enable()
+        await asyncio.sleep(15)
+
+
+@routes.post('/api/v1alpha/instances/activate')
+# @rest_authenticated_users_only
+async def activate_worker(request):
+    return await asyncio.shield(app['driver'].activate_worker(request))
+
+
+@routes.post('/api/v1alpha/instances/deactivate')
+# @rest_authenticated_users_only
+async def deactivate_worker(request):
+    return await asyncio.shield(app['driver'].deactivate_worker(request))
+
+
+@routes.post('/api/v1alpha/instances/pod_complete')
+# @rest_authenticated_users_only
+async def pod_complete(request):
+    return await asyncio.shield(app['driver'].pod_complete(request))
+
+
+setup_aiohttp_jinja2(app, 'batch')
+
+setup_common_static_routes(routes)
+
 app.add_routes(routes)
-deploy_config = get_deploy_config()
-web.run_app(deploy_config.prefix_application(app, 'batch2'), host='0.0.0.0', port=5000)
+
+app.router.add_get("/metrics", server_stats)
+
+
+async def on_startup(app):
+    profile = cProfile.Profile()
+    profile.enable()
+    app['profile'] = profile
+
+    pool = concurrent.futures.ThreadPoolExecutor()
+    k8s = K8s(pool, KUBERNETES_TIMEOUT_IN_SECONDS, BATCH_NAMESPACE, v1)
+
+    userinfo = await async_get_userinfo()
+    bucket_name = userinfo['bucket_name']
+
+    driver = Driver(k8s, bucket_name)
+
+    app['blocking_pool'] = pool
+    app['driver'] = driver
+    app['log_store'] = LogStore(pool, INSTANCE_ID, bucket_name)
+
+    await driver.initialize()
+    await refresh_pods()  # this is really slow for large N
+
+    asyncio.ensure_future(driver.run())
+    # we need a polling event loop in case a delete happens before a create job, but this is too slow
+    # we also need a polling loop in case pod creation fails
+    # asyncio.ensure_future(polling_event_loop())
+    asyncio.ensure_future(driver_event_loop())
+    asyncio.ensure_future(db_cleanup_event_loop())
```

Comment:
I think we should probably capture these tasks into the `app` and then `cancel` them in `on_cleanup`. Otherwise, I suspect the python process will hang until it's forcefully terminated.